machine/models/baseModel.py

```diff
@@ -0,0 +1,49 @@
+import torch.nn as nn
+import torch.nn.functional as F
+
+class Model(nn.Module):
+    """ 
+    Abstract base class for models.
+
+    Args:
+        encoder_module (baseRNN) :  module that encodes inputs
+        decoder_module (baseRNN, optional):   module to decode encoded inputs
+        decode_function (func, optional): function to generate symbols from output hidden states (default: F.log_softmax)
+
+    """
+
+    def __init__(self, encoder_module, decoder_module=None, decode_function=F.log_softmax):
+        super(Model, self).__init__()
+        self.encoder_module = encoder_module
+        self.decoder_module = decoder_module
+        self.decode_function = decode_function
+
+    def flatten_parameters(self):
+        self.encoder_module.rnn.flatten_parameters()
+        self.decoder_module.rnn.flatten_parameters()
+
+    def forward(self, inputs, input_lengths=None, targets=None,
+                teacher_forcing_ratio=0):
+        """
+        Inputs: inputs, input_lengths, targets, teacher_forcing_ratio
+            - **inputs** (list, option): list of sequences, whose length is the batch size and within which
+              each sequence is a list of token IDs. This information is passed to the encoder module.
+            - **input_lengths** (list of int, optional): A list that contains the lengths of sequences
+                in the mini-batch, it must be provided when using variable length RNN (default: `None`)
+            - **targets** (list, optional): list of sequences, whose length is the batch size and within which
+              each sequence is a list of token IDs. This information is forwarded to the decoder.
+            - **teacher_forcing_ratio** (int, optional): The probability that teacher forcing will be used. A random number
+              is drawn uniformly from 0-1 for every decoding token, and if the sample is smaller than the given value,
+              teacher forcing would be used (default is 0)
+
+        Outputs: decoder_outputs, decoder_hidden, ret_dict
+            - **outputs** (batch): batch-length list of tensors with size (max_length, hidden_size) containing the
```

Comment:
`batch` --> `list of nn.Tensors` ?

EDIT: I just realized you just give the sizes for outputs , why not the type ? `outputs` is a list `decoder_hidden` is a tensor but this doesn't appear inside the parentheses